protocol/contracts/templegold/SpiceAuction.sol

```diff
@@ -0,0 +1,253 @@
+        delete auctionConfigs[id];
+        /// @dev `_currentEpochId` is only updated when auction starts
+        emit AuctionConfigRemoved(id);
+    }
+
+    /**
+     * @notice Start auction. Checks caller is set config starter. Address zero for anyone to call
+     */
+    function startAuction() external override {
+        uint256 epochId = _currentEpochId;
+        /// @dev config is always set for next auction
+        /// @notice Configuration is set before auctions so configId = currentEpochId + 1;
+        SpiceAuctionConfig storage config = auctionConfigs[epochId+1];
+        if (config.duration == 0) { revert CannotStartAuction(); }
+        /// @notice only starter
+        if (config.starter != address(0) && msg.sender != config.starter) { revert CommonEventsAndErrors.InvalidAccess(); }
+        /// @notice enough wait period since last auction
+        if (epochId > 0) {
+            /// @dev `_currentEpochId` is still last epoch
+            EpochInfo memory lastEpochInfo = epochs[epochId];
+            /// use waitperiod from last auction config
+            uint64 _waitPeriod = auctionConfigs[epochId].waitPeriod;
+            if (lastEpochInfo.endTime + _waitPeriod > block.timestamp) { revert CannotStartAuction(); }
+        } else {
+            /// For first auction
+            if (_deployTimestamp + config.waitPeriod > block.timestamp) { revert CannotStartAuction(); }
+        }
+        address auctionToken = _getAuctionTokenForEpochStorage(config);
+        uint256 totalAuctionTokenAllocation = _totalAuctionTokenAllocation[auctionToken];
+        uint256 balance = IERC20(auctionToken).balanceOf(address(this));
+        uint256 epochAuctionTokenAmount = balance - (totalAuctionTokenAllocation - _claimedAuctionTokens[auctionToken]);
+        if (config.activationMode == ActivationMode.AUCTION_TOKEN_BALANCE) {
+            if (config.minimumDistributedAuctionToken == 0) { revert MissingAuctionTokenConfig(); }
+        }
+        if (epochAuctionTokenAmount < config.minimumDistributedAuctionToken) { revert NotEnoughAuctionTokens(); }
+        // epoch start settings
+        // now update currentEpochId
+        epochId = _currentEpochId = _currentEpochId + 1;
+        EpochInfo storage info = epochs[epochId];
+        uint64 startTime = info.startTime = uint64(block.timestamp) + config.startCooldown;
+        uint64 endTime = info.endTime = uint64(block.timestamp) + config.duration;
+        info.totalAuctionTokenAmount = epochAuctionTokenAmount;
+
+        // Keep track of total allocation auction tokens per epoch
+        _totalAuctionTokenAllocation[auctionToken] = totalAuctionTokenAllocation + epochAuctionTokenAmount;
+
+        emit AuctionStarted(epochId, msg.sender, startTime, endTime, epochAuctionTokenAmount);
+    }
+
+    /**
+     * @notice Bid using `bidToken` for `auctionToken`
+     * @param amount Amount of `bidToken` to bid
+     */
+    function bid(uint256 amount) external virtual override {
```

Comment:
It's currently not possible to cancel a bid. Just checking this is intended (which is fine, just checking)

Probably want to add to the docstring here for that.

Reply:
yes can't cancel a bid. had code before to cancel and withdraw bids, but we later agreed it opens a way for whales to gamify the auctions for personal gains.